File: android/src/main/kotlin/twilio/flutter/twilio_programmable_chat/PaginatorManager.kt
package twilio.flutter.twilio_programmable_chat

import com.twilio.chat.Paginator
import java.util.UUID

object PaginatorManager {
    private val paginators: HashMap<String, Paginator<Any>> = hashMapOf()

    @Suppress("UNCHECKED_CAST")
    fun <T> setPaginator(paginator: Paginator<T>): String {
        val pageId = UUID.randomUUID().toString()
        paginators[pageId] = paginator as Paginator<Any>
        return pageId
    }

    @Suppress("UNCHECKED_CAST")
    fun <T> getPaginator(pageId: String): Paginator<T> {
        if (!paginators.containsKey(pageId)) {
            throw Exception("Could not find paginator with pageId '$pageId'")
        }
        return paginators[pageId] as Paginator<T>
    }
}

File: ios/Classes/PaginatorManager.swift
import Foundation
import TwilioChatClient

public class PaginatorManager {
    internal static var paginators = [String: Any]()

    public static func setPaginator(_ paginator: Any) -> String {
        let pageId = UUID.init().uuidString
        paginators[pageId] = paginator
        return pageId
    }

    public static func getPaginator<T>(_ pageId: String) throws -> T {
        if !paginators.keys.contains(pageId) {
            throw PaginatorManagerError("Could not find paginator with pageId '\(pageId)'")
        } else {
            if let paginator = paginators[pageId] as? T {
                return paginator
            } else {
                throw PaginatorManagerError("Error converting '\(String(describing: paginators[pageId]))' to \(T.self)")
            }
        }
    }
}

struct PaginatorManagerError: Error {
    let error: String

    init(_ errorString: String) {
        self.error = errorString
    }
}
